File: swift/0115-distinct-subsequences.swift
/**
 * Question Link: https://leetcode.com/problems/distinct-subsequences/
 */

 class Solution {
    func numDistinct(_ s: String, _ t: String) -> Int {
        let s = Array(s), t = Array(t)
        let m = s.count, n = t.count
        var dp = Array(repeating: 0.0, count: n)
        for i in stride(from: m - 1, to: -1, by: -1) {
            var prev = 1.0
            for j in stride(from: n - 1, to: -1, by: -1) {
                let oldDPJ = dp[j]
                if s[i] == t[j] {
                    dp[j] += prev
                }
                prev = oldDPJ
            }
        }

        return Int(dp[0])
    }
}

File: java/0115-distinct-subsequences.java
import java.util.Arrays;

class Solution {
    public int numDistinct(String s, String t) {
        int n = s.length();
        int m = t.length();
       
        
        int[] dp = new int[m + 1];
        dp[0] = 1; 
       
        for (int i = 0; i < n; i++) {
            for (int j = m; j > 0; j--) {
                if (s.charAt(i) == t.charAt(j - 1)) {
                    dp[j] += dp[j - 1]; 
                }
            }
        }
       
        return dp[m];
    }
}
